File: packages/less/src/less/tree/condition.test.js
import { describe, it, expect, vi } from 'vitest';
import Condition from './condition';
import Node from './node'; // Import Node for Node.compare

// Mock Node for operands - simplified version for testing Condition logic
class MockNode extends Node {
    constructor(value, type = 'Mock') {
        super();
        this._value = value;
        this.type = type;
    }

    eval() {
        // Return the mock node itself for comparison logic
        return this;
    }

    // Simple compare for testing purposes, mimicking Node.compare logic for basic values
    compare(other) {
        if (other.type !== this.type) {
            return undefined;
        }
        if (this._value < other._value) {
            return -1;
        } else if (this._value > other._value) {
            return 1;
        } else {
            return 0;
        }
    }

    get value() {
        // Needed for Node.compare fallback if types match but compare isn't called directly
        return this._value;
    }
}

// Mock nodes evaluating to boolean true/false for logical operators
const TRUE_NODE = { eval: () => true };
const FALSE_NODE = { eval: () => false };

// Mock nodes evaluating to comparable numeric values
const ONE = new MockNode(1, 'Dimension');
const TWO = new MockNode(2, 'Dimension');
const ANOTHER_TWO = new MockNode(2, 'Dimension'); // Same value, different instance

// Mock nodes evaluating to comparable string values
const A_STR = new MockNode('a', 'Quoted');
const B_STR = new MockNode('b', 'Quoted');
const ANOTHER_A_STR = new MockNode('a', 'Quoted');

// Mock nodes of different types
const ONE_KEYWORD = new MockNode(1, 'Keyword');

describe('Condition', () => {
    describe('eval', () => {
        // --- Logical Operators ---
        describe('and', () => {
            it('should evaluate true and true to true', () => {
                const condition = new Condition('and', TRUE_NODE, TRUE_NODE);
                expect(condition.eval({})).toBe(true);
            });
            it('should evaluate true and false to false', () => {
                const condition = new Condition('and', TRUE_NODE, FALSE_NODE);
                expect(condition.eval({})).toBe(false);
            });
            it('should evaluate false and true to false', () => {
                const condition = new Condition('and', FALSE_NODE, TRUE_NODE);
                expect(condition.eval({})).toBe(false);
            });
            it('should evaluate false and false to false', () => {
                const condition = new Condition('and', FALSE_NODE, FALSE_NODE);
                expect(condition.eval({})).toBe(false);
            });
            it('should handle negation', () => {
                const condition = new Condition(
                    'and',
                    TRUE_NODE,
                    TRUE_NODE,
                    0,
                    true
                ); // negate = true
                expect(condition.eval({})).toBe(false); // !(true && true) -> false
            });
        });

        describe('or', () => {
            it('should evaluate true or true to true', () => {
                const condition = new Condition('or', TRUE_NODE, TRUE_NODE);
                expect(condition.eval({})).toBe(true);
            });
            it('should evaluate true or false to true', () => {
                const condition = new Condition('or', TRUE_NODE, FALSE_NODE);
                expect(condition.eval({})).toBe(true);
            });
            it('should evaluate false or true to true', () => {
                const condition = new Condition('or', FALSE_NODE, TRUE_NODE);
                expect(condition.eval({})).toBe(true);
            });
            it('should evaluate false or false to false', () => {
                const condition = new Condition('or', FALSE_NODE, FALSE_NODE);
                expect(condition.eval({})).toBe(false);
            });
            it('should handle negation', () => {
                const condition = new Condition(
                    'or',
                    FALSE_NODE,
                    FALSE_NODE,
                    0,
                    true
                ); // negate = true
                expect(condition.eval({})).toBe(true); // !(false || false) -> true
            });
        });

        // --- Comparison Operators ---
        // Using numeric MockNodes (ONE, TWO, ANOTHER_TWO)
        describe('Comparison Operators (Numbers)', () => {
            it('< : 1 < 2 should be true', () => {
                const condition = new Condition('<', ONE, TWO);
                expect(condition.eval({})).toBe(true);
            });
            it('< : 2 < 1 should be false', () => {
                const condition = new Condition('<', TWO, ONE);
                expect(condition.eval({})).toBe(false);
            });
            it('< : 2 < 2 should be false', () => {
                const condition = new Condition('<', TWO, ANOTHER_TWO);
                expect(condition.eval({})).toBe(false);
            });
            it('<= : 1 <= 2 should be true', () => {
                const condition = new Condition('<=', ONE, TWO);
                expect(condition.eval({})).toBe(true);
            });
            it('<= : 2 <= 1 should be false', () => {
                const condition = new Condition('<=', TWO, ONE);
                expect(condition.eval({})).toBe(false);
            });
            it('<= : 2 <= 2 should be true', () => {
                const condition = new Condition('<=', TWO, ANOTHER_TWO);
                expect(condition.eval({})).toBe(true);
            });
            it('=< : 1 =< 2 should be true', () => {
                const condition = new Condition('=<', ONE, TWO);
                expect(condition.eval({})).toBe(true);
            });
            it('=< : 2 =< 1 should be false', () => {
                const condition = new Condition('=<', TWO, ONE);
                expect(condition.eval({})).toBe(false);
            });
            it('=< : 2 =< 2 should be true', () => {
                const condition = new Condition('=<', TWO, ANOTHER_TWO);
                expect(condition.eval({})).toBe(true);
            });
            it('= : 1 = 2 should be false', () => {
                const condition = new Condition('=', ONE, TWO);
                expect(condition.eval({})).toBe(false);
            });
            it('= : 2 = 2 should be true', () => {
                const condition = new Condition('=', TWO, ANOTHER_TWO);
                expect(condition.eval({})).toBe(true);
            });
            it('>= : 1 >= 2 should be false', () => {
                const condition = new Condition('>=', ONE, TWO);
                expect(condition.eval({})).toBe(false);
            });
            it('>= : 2 >= 1 should be true', () => {
                const condition = new Condition('>=', TWO, ONE);
                expect(condition.eval({})).toBe(true);
            });
            it('>= : 2 >= 2 should be true', () => {
                const condition = new Condition('>=', TWO, ANOTHER_TWO);
                expect(condition.eval({})).toBe(true);
            });
            it('> : 1 > 2 should be false', () => {
                const condition = new Condition('>', ONE, TWO);
                expect(condition.eval({})).toBe(false);
            });
            it('> : 2 > 1 should be true', () => {
                const condition = new Condition('>', TWO, ONE);
                expect(condition.eval({})).toBe(true);
            });
            it('> : 2 > 2 should be false', () => {
                const condition = new Condition('>', TWO, ANOTHER_TWO);
                expect(condition.eval({})).toBe(false);
            });
            it('should handle negation for comparison', () => {
                const condition = new Condition('<', ONE, TWO, 0, true); // negate = true
                expect(condition.eval({})).toBe(false); // !(1 < 2) -> false
            });
            it('should return false for comparison of incompatible types', () => {
                // Node.compare returns undefined for different types, resulting in false
                const condition = new Condition('=', ONE, ONE_KEYWORD);
                expect(condition.eval({})).toBe(false);
            });
        });

        // Using string MockNodes (A_STR, B_STR, ANOTHER_A_STR)
        // Note: Node.compare for Quoted falls back to value comparison
        describe('Comparison Operators (Strings)', () => {
            it('< : "a" < "b" should be true', () => {
                const condition = new Condition('<', A_STR, B_STR);
                expect(condition.eval({})).toBe(true);
            });
            it('= : "a" = "a" should be true', () => {
                const condition = new Condition('=', A_STR, ANOTHER_A_STR);
                expect(condition.eval({})).toBe(true);
            });
            it('> : "b" > "a" should be true', () => {
                const condition = new Condition('>', B_STR, A_STR);
                expect(condition.eval({})).toBe(true);
            });
            it('>= : "a" >= "a" should be true', () => {
                const condition = new Condition('>=', A_STR, ANOTHER_A_STR);
                expect(condition.eval({})).toBe(true);
            });
            it('<= : "a" <= "a" should be true', () => {
                const condition = new Condition('<=', A_STR, ANOTHER_A_STR);
                expect(condition.eval({})).toBe(true);
            });
            it('should handle empty strings correctly', () => {
                const empty1 = new MockNode('', 'Quoted');
                const empty2 = new MockNode('', 'Quoted');
                const nonEmpty = new MockNode('a', 'Quoted');

                const condition1 = new Condition('=', empty1, empty2);
                const condition2 = new Condition('=', empty1, nonEmpty);

                expect(condition1.eval({})).toBe(true);
                expect(condition2.eval({})).toBe(false);
            });
            it('should handle special characters in strings', () => {
                const special1 = new MockNode('a@#$%', 'Quoted');
                const special2 = new MockNode('a@#$%', 'Quoted');
                const different = new MockNode('b@#$%', 'Quoted');

                const condition1 = new Condition('=', special1, special2);
                const condition2 = new Condition('=', special1, different);

                expect(condition1.eval({})).toBe(true);
                expect(condition2.eval({})).toBe(false);
            });
        });

        describe('Edge Cases and Special Values', () => {
            const nullNode = new MockNode(null, 'Null');
            const undefinedNode = new MockNode(undefined, 'Undefined');
            const largeNumber = new MockNode(
                Number.MAX_SAFE_INTEGER,
                'Dimension'
            );
            const negativeNumber = new MockNode(-1, 'Dimension');
            const zero = new MockNode(0, 'Dimension');

            it('should handle null values', () => {
                const condition = new Condition('=', nullNode, nullNode);
                expect(condition.eval({})).toBe(true); // Same type null values are equal
            });

            it('should handle undefined values', () => {
                const condition = new Condition(
                    '=',
                    undefinedNode,
                    undefinedNode
                );
                expect(condition.eval({})).toBe(true); // Same type undefined values are equal
            });

            it('should handle large numbers', () => {
                const condition = new Condition('>', largeNumber, zero);
                expect(condition.eval({})).toBe(true);
            });

            it('should handle negative numbers', () => {
                const condition = new Condition('<', negativeNumber, zero);
                expect(condition.eval({})).toBe(true);
            });
        });

        describe('Type Coercion and Compatibility', () => {
            const numberNode = new MockNode(1, 'Dimension');
            const stringNumber = new MockNode('1', 'Quoted');
            const keywordNode = new MockNode('test', 'Keyword');
            const colorNode = new MockNode('#000000', 'Color');

            it('should handle string-number comparisons', () => {
                const condition = new Condition('=', numberNode, stringNumber);
                expect(condition.eval({})).toBe(false); // Different types should return false
            });

            it('should handle keyword comparisons', () => {
                const condition = new Condition('=', keywordNode, keywordNode);
                expect(condition.eval({})).toBe(true); // Same type keywords are equal
            });

            it('should handle color comparisons', () => {
                const condition = new Condition('=', colorNode, colorNode);
                expect(condition.eval({})).toBe(true); // Same type colors are equal
            });
        });

        describe('Error Cases', () => {
            it('should handle invalid operators gracefully', () => {
                const condition = new Condition('invalid', ONE, TWO);
                expect(condition.eval({})).toBe(false); // Invalid operators should return false
            });

            it('should handle malformed input gracefully', () => {
                const condition = new Condition('=', null, TWO);
                expect(() => condition.eval({})).toThrow();
            });
        });
    });

    describe('accept', () => {
        it('should call visitor.visit on lvalue and rvalue', () => {
            const lvalue = new MockNode(1);
            const rvalue = new MockNode(2);
            const condition = new Condition('>', lvalue, rvalue);

            const mockVisitor = {
                visit: vi.fn((node) => {
                    // Return a slightly modified node for verification
                    return new MockNode(node.value + 10, node.type);
                })
            };

            condition.accept(mockVisitor);

            // Check if visit was called twice
            expect(mockVisitor.visit).toHaveBeenCalledTimes(2);

            // Check if visit was called with the original nodes
            expect(mockVisitor.visit).toHaveBeenCalledWith(lvalue);
            expect(mockVisitor.visit).toHaveBeenCalledWith(rvalue);

            // Check if lvalue and rvalue were updated with the results from visit
            expect(condition.lvalue.value).toBe(11); // 1 + 10
            expect(condition.rvalue.value).toBe(12); // 2 + 10
        });
    });

    describe('properties', () => {
        it('should store constructor properties correctly', () => {
            const lvalue = new MockNode(1);
            const rvalue = new MockNode(2);
            const condition = new Condition(' > ', lvalue, rvalue, 5, true);

            expect(condition.op).toBe('>'); // Check trimming
            expect(condition.lvalue).toBe(lvalue);
            expect(condition.rvalue).toBe(rvalue);
            expect(condition._index).toBe(5);
            expect(condition.negate).toBe(true);
            expect(condition.type).toBe('Condition');
        });
    });
});

File: packages/less/src/less/tree/condition_test.go
package tree

import (
	"reflect"
	"testing"
)

// MockNode is a simplified version of Node for testing Condition logic
type MockNode struct {
	*Node
	Value any
	Type  string
}

// NewMockNode creates a new MockNode for testing
func NewMockNode(value any, nodeType string) *MockNode {
	node := NewNode()
	mockNode := &MockNode{
		Node:  node,
		Value: value,
		Type:  nodeType,
	}
	// Set the mockNode as the Node's Value for proper comparison
	node.Value = mockNode
	return mockNode
}

// Eval returns the MockNode itself for comparison logic
func (m *MockNode) Eval(context any) any {
	return m.Node
}

// Compare implements comparison between MockNodes
func (m *MockNode) Compare(other *Node) int {
	if other == nil || other.Value == nil {
		return 0
	}
	
	otherMock, ok := other.Value.(*MockNode)
	if !ok {
		return 0 // Different value types
	}
	
	// Different node types should return 0 (incomparable)
	if otherMock.Type != m.Type {
		return 0
	}

	switch v := m.Value.(type) {
	case int:
		if otherVal, ok := otherMock.Value.(int); ok {
			if v < otherVal {
				return -1
			} else if v > otherVal {
				return 1
			}
			return 0
		}
	case string:
		if otherVal, ok := otherMock.Value.(string); ok {
			if v < otherVal {
				return -1
			} else if v > otherVal {
				return 1
			}
			return 0
		}
	}
	
	return 0
}

// Special method to indicate the node type - this helps the Node.Compare method
// identify different types correctly
func (m *MockNode) String() string {
	return m.Type
}

// Test helpers for logical operations
type BoolEvaluator struct {
	Value bool
}

func (b *BoolEvaluator) Eval(context any) any {
	return b.Value
}

func TestCondition_EvalLogicalOperators(t *testing.T) {
	// Create boolean evaluators
	trueNode := &BoolEvaluator{Value: true}
	falseNode := &BoolEvaluator{Value: false}

	tests := []struct {
		name     string
		op       string
		lvalue   any
		rvalue   any
		negate   bool
		expected bool
	}{
		{"true and true", "and", trueNode, trueNode, false, true},
		{"true and false", "and", trueNode, falseNode, false, false},
		{"false and true", "and", falseNode, trueNode, false, false},
		{"false and false", "and", falseNode, falseNode, false, false},
		{"negated (true and true)", "and", trueNode, trueNode, true, false},
		
		{"true or true", "or", trueNode, trueNode, false, true},
		{"true or false", "or", trueNode, falseNode, false, true},
		{"false or true", "or", falseNode, trueNode, false, true},
		{"false or false", "or", falseNode, falseNode, false, false},
		{"negated (false or false)", "or", falseNode, falseNode, true, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			condition := NewCondition(tt.op, tt.lvalue, tt.rvalue, 0, tt.negate)
			result := condition.Eval(nil)
			if result != tt.expected {
				t.Errorf("Expected %v, got %v", tt.expected, result)
			}
		})
	}
}

func TestCondition_EvalComparisonOperators(t *testing.T) {
	// Numeric test values
	one := NewMockNode(1, "Dimension")
	two := NewMockNode(2, "Dimension")
	anotherTwo := NewMockNode(2, "Dimension")
	
	// String test values
	aStr := NewMockNode("a", "Quoted")
	bStr := NewMockNode("b", "Quoted")
	anotherAStr := NewMockNode("a", "Quoted")
	
	// Different type value
	oneKeyword := NewMockNode(1, "Keyword")

	// Mock comparisons that should pass
	mockCompareTrue := []struct {
		name   string
		op     string
		lvalue *MockNode
		rvalue *MockNode
	}{
		{"1 < 2", "<", one, two},
		{"1 <= 2", "<=", one, two},
		{"2 <= 2", "<=", two, anotherTwo},
		{"1 =< 2", "=<", one, two},
		{"2 =< 2", "=<", two, anotherTwo},
		{"2 = 2", "=", two, anotherTwo},
		{"2 >= 1", ">=", two, one},
		{"2 >= 2", ">=", two, anotherTwo},
		{"2 > 1", ">", two, one},
		{"a < b", "<", aStr, bStr},
		{"a = a", "=", aStr, anotherAStr},
		{"b > a", ">", bStr, aStr},
	}

	// Mock comparisons that should fail
	mockCompareFalse := []struct {
		name   string
		op     string
		lvalue *MockNode
		rvalue *MockNode
	}{
		{"2 < 1", "<", two, one},
		{"2 < 2", "<", two, anotherTwo},
		{"2 <= 1", "<=", two, one},
		{"2 =< 1", "=<", two, one},
		{"1 = 2", "=", one, two},
		{"1 >= 2", ">=", one, two},
		{"1 > 2", ">", one, two},
		{"2 > 2", ">", two, anotherTwo},
		{"1(dimension) = 1(keyword)", "=", one, oneKeyword},
	}

	// Test successful comparisons
	for _, tt := range mockCompareTrue {
		t.Run(tt.name, func(t *testing.T) {
			condition := NewCondition(tt.op, tt.lvalue.Node, tt.rvalue.Node, 0, false)
			result := condition.Eval(nil)
			if result != true {
				t.Errorf("Expected true for %s, got %v", tt.name, result)
			}
		})
	}

	// Test failed comparisons
	for _, tt := range mockCompareFalse {
		t.Run(tt.name, func(t *testing.T) {
			condition := NewCondition(tt.op, tt.lvalue.Node, tt.rvalue.Node, 0, false)
			result := condition.Eval(nil)
			if result != false {
				t.Errorf("Expected false for %s, got %v", tt.name, result)
			}
		})
	}

	// Test negation
	t.Run("!(1 < 2)", func(t *testing.T) {
		condition := NewCondition("<", one.Node, two.Node, 0, true)
		result := condition.Eval(nil)
		if result != false {
			t.Errorf("Expected false for negated '1 < 2', got %v", result)
		}
	})
}

func TestDifferentTypes(t *testing.T) {
	// Test that different MockNode types are not considered equal
	dimension := NewMockNode(1, "Dimension")
	keyword := NewMockNode(1, "Keyword")

	// Verify the types are actually different in our objects
	if reflect.TypeOf(dimension) == reflect.TypeOf(keyword) {
		t.Logf("Types correctly match: both are %T", dimension)
	}
	
	if dimension.Type == keyword.Type {
		t.Errorf("Types should be different: %s vs %s", dimension.Type, keyword.Type)
	}

	// Direct comparison isn't reliable since Compare might return 0
	// even for different types - we rely on Condition to handle that
	result := Compare(dimension.Node, keyword.Node)
	t.Logf("Compare result: %d", result)

	// The important part is that Condition evaluates correctly
	condition := NewCondition("=", dimension.Node, keyword.Node, 0, false)
	condResult := condition.Eval(nil)
	if condResult {
		t.Error("Condition with different types should evaluate to false for equality")
	} else {
		t.Logf("Condition correctly evaluated to false for different types")
	}
}

func TestCondition_Accept(t *testing.T) {
	// Create a custom visitor to update the mockNode values
	visitCount := 0
	visitor := &testVisitor{
		visitFunc: func(node any) any {
			visitCount++
			if nodeVal, ok := node.(*Node); ok {
				if mockNode, ok := nodeVal.Value.(*MockNode); ok {
					if val, ok := mockNode.Value.(int); ok {
						newMock := NewMockNode(val+10, mockNode.Type)
						return newMock.Node
					}
				}
			}
			return node
		},
	}

	// Create and test the condition
	lvalue := NewMockNode(1, "Dimension")
	rvalue := NewMockNode(2, "Dimension")
	condition := NewCondition(">", lvalue.Node, rvalue.Node, 0, false)
	
	// Call Accept with the visitor
	condition.Accept(visitor)
	
	// Check visit count
	if visitCount != 2 {
		t.Errorf("Expected 2 visits, got %d", visitCount)
	}
	
	// Verify the values were updated
	lnode, ok := condition.Lvalue.(*Node)
	if !ok {
		t.Fatalf("Expected lvalue to be *Node, got %T", condition.Lvalue)
	}
	
	lmock, ok := lnode.Value.(*MockNode)
	if !ok {
		t.Fatalf("Expected lnode.Value to be *MockNode, got %T", lnode.Value)
	}
	
	if lval, ok := lmock.Value.(int); !ok || lval != 11 {
		t.Errorf("Expected lvalue to have value 11, got %v", lmock.Value)
	}
	
	rnode, ok := condition.Rvalue.(*Node)
	if !ok {
		t.Fatalf("Expected rvalue to be *Node, got %T", condition.Rvalue)
	}
	
	rmock, ok := rnode.Value.(*MockNode)
	if !ok {
		t.Fatalf("Expected rnode.Value to be *MockNode, got %T", rnode.Value)
	}
	
	if rval, ok := rmock.Value.(int); !ok || rval != 12 {
		t.Errorf("Expected rvalue to have value 12, got %v", rmock.Value)
	}
}

func TestCondition_Properties(t *testing.T) {
	// Verify constructor properties are stored correctly
	lvalue := NewMockNode(1, "Dimension")
	rvalue := NewMockNode(2, "Dimension")
	condition := NewCondition(" > ", lvalue.Node, rvalue.Node, 5, true)

	if condition.Op != ">" {
		t.Errorf("Expected trimmed op '>', got '%s'", condition.Op)
	}

	if condition.Lvalue != lvalue.Node {
		t.Errorf("Expected lvalue to be %v, got %v", lvalue.Node, condition.Lvalue)
	}

	if condition.Rvalue != rvalue.Node {
		t.Errorf("Expected rvalue to be %v, got %v", rvalue.Node, condition.Rvalue)
	}

	if condition.Index != 5 {
		t.Errorf("Expected index to be 5, got %d", condition.Index)
	}

	if !condition.Negate {
		t.Errorf("Expected negate to be true, got %v", condition.Negate)
	}
} 